File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/addressspace/nodes/UaNode.java
package org.opcfoundation.webserver.addressspace.nodes;

import org.eclipse.milo.opcua.stack.core.AttributeId;
import org.eclipse.milo.opcua.stack.core.NodeIds;
import org.eclipse.milo.opcua.stack.core.types.builtin.*;
import org.eclipse.milo.opcua.stack.core.types.builtin.unsigned.UInteger;
import org.eclipse.milo.opcua.stack.core.types.enumerated.BrowseDirection;
import org.eclipse.milo.opcua.stack.core.types.enumerated.NodeClass;
import org.springframework.lang.Nullable;
import org.opcfoundation.webserver.addressspace.nodes.builtin.UaReferenceTypes;

import java.util.ArrayList;
import java.util.List;

public abstract class UaNode {
    private final NodeId nodeId;
    protected String browseName;
    protected LocalizedText displayName;
    protected UInteger writeMask;
    protected LocalizedText description;
    protected List<UaReference> references;

    public UaNode(
            NodeId nodeId,
            String browseName,
            LocalizedText displayName)
    {
        this.nodeId = nodeId;
        this.browseName = browseName;
        this.displayName = displayName;
        this.writeMask = UInteger.valueOf(0);
        this.description = LocalizedText.NULL_VALUE;
        this.references = new ArrayList<>();
    }

    public NodeId nodeId() {
        return nodeId;
    }

    public abstract NodeClass nodeClass();

    public boolean isDefinitionNode()
    {
        return (NodeClass.ObjectType == nodeClass() ||
                NodeClass.VariableType == nodeClass() ||
                NodeClass.DataType == nodeClass() ||
                NodeClass.ReferenceType == nodeClass());
    }

    public boolean isInstanceNode()
    {
        return (NodeClass.Object == nodeClass() ||
                NodeClass.Variable == nodeClass() ||
                NodeClass.Method == nodeClass());
    }

    public String browseName() {
        return browseName;
    }

    public void browseName(String browseName) {
        this.browseName = browseName;
    }

    public LocalizedText displayName() {
        return displayName;
    }

    public void setDisplayName(LocalizedText displayName) {
        this.displayName = displayName;
    }

    public UInteger writeMask() {
        return writeMask;
    }

    public void setWriteMask(UInteger writeMask) {
        this.writeMask = writeMask;
    }

    public LocalizedText description() {
        return description;
    }

    public void setDescription(LocalizedText description) {
        this.description = description;
    }

    public Variant getAttribute(int attributeId)
    {
        Variant value = Variant.NULL_VALUE;

        if (AttributeId.NodeId.id() == attributeId) {
            value = new Variant(nodeId);
        } else if (AttributeId.NodeClass.id() == attributeId) {
            value = new Variant(nodeClass().getValue());
        } else if (AttributeId.BrowseName.id() == attributeId) {
            value = new Variant(new QualifiedName(0,browseName));
        } else if (AttributeId.DisplayName.id() == attributeId) {
            value = new Variant(displayName);
        } else if (AttributeId.Description.id() == attributeId) {
            if (null != description) value = new Variant(description);
        } else if (AttributeId.WriteMask.id() == attributeId) {
            value = new Variant(writeMask);
        } else if (AttributeId.UserWriteMask.id() == attributeId) {
            value = new Variant(0);
        }

        return value;
    }

    public List<UaReference> getReferences(BrowseDirection direction)
    {
        List<UaReference> ret = new ArrayList<>();
        if (null == this.references) return ret;

        for (UaReference item: references)
        {
            if ((BrowseDirection.Forward == direction && !item.isForward()) ||
                (BrowseDirection.Inverse == direction && item.isForward())) continue;

            ret.add(item);
        }

        return ret;
    }

    public List<UaInstanceNode> getMembers()
    {
        List<UaReference> referenceList = getReferences(BrowseDirection.Forward);
        ArrayList<UaInstanceNode> members = new ArrayList<>();

        for (UaReference item: referenceList)
        {
            if (!item.linkedNode().isInstanceNode()) continue;
            if (!item.reference().isSubtypeOf(NodeIds.Aggregates)) continue;
            members.add((UaInstanceNode)item.linkedNode());
        }

        return members;
    }

    public boolean hasMember()
    {
        for (UaReference item: references)
        {
            if (item.isForward() &&
                    item.linkedNode().isInstanceNode() &&
                    item.reference().isSubtypeOf(NodeIds.Aggregates)) return true;
        }

        return false;
    }

    public @Nullable UaInstanceNode getMember(String path)
    {
        for (UaReference item : references)
        {
            if (item.isForward() &&
                    item.linkedNode().browseName().equals(path) &&
                    item.linkedNode().isInstanceNode() &&
                    item.reference().isSubtypeOf(NodeIds.Aggregates)) return (UaInstanceNode)item.linkedNode();
        }

        return null;
    }

    protected void addReference(UaReference reference)
    {
        references.add(reference);
    }

    protected void addMemberNode(UaInstanceNode member)
    {
        UaReferenceType reference = UaReferenceTypes.HasComponent;

        if (NodeClass.Variable == member.nodeClass()) {
            if (((UaVariable)member).typeDefinition().nodeId().equals(NodeIds.PropertyType))
            {
                reference = UaReferenceTypes.HasProperty;
            }
        }

        addReference(new UaReference(member, reference, true));
        member.addReference(new UaReference(this,reference, false));
    }
}

File: opcua-webservice-node/source/src/server/addressspace/nodes/UaNode.ts
import { Attributes, BrowseDirection, NodeClass } from "opcua-webapi";
import { ReferenceTypeIds, UaLocalizedText, UaNodeId, UaVariant, UaVariantType, VariableTypeIds } from "opcua-webapi-ts";
import { UaReference } from "./UaReference";
import type { UaInstanceNode } from "./UaInstanceNode";
import type { UaVariable } from "./UaVariable";

function getUaReferenceTypes(): typeof import("./builtin/UaReferenceTypes").UaReferenceTypes {
    return require("./builtin/UaReferenceTypes").UaReferenceTypes;
}

export abstract class UaNode {
    private readonly _nodeId: UaNodeId;
    protected _browseName: string;
    protected _displayName: UaLocalizedText;
    protected _writeMask: number;
    protected _description: UaLocalizedText;
    protected _references: Array<UaReference>;

    constructor(
        nodeId: UaNodeId,
        browseName: string,
        displayName: UaLocalizedText) {
        this._nodeId = nodeId;
        this._browseName = browseName;
        this._displayName = displayName;
        this._writeMask = 0;
        this._description = UaLocalizedText.nullText;
        this._references = [];
    }

    get nodeId(): UaNodeId {
        return this._nodeId;
    }

    abstract get nodeClass(): NodeClass;

    public isDefinitionNode(): boolean {
        return (NodeClass.ObjectType === this.nodeClass ||
                NodeClass.VariableType === this.nodeClass ||
                NodeClass.DataType === this.nodeClass ||
                NodeClass.ReferenceType === this.nodeClass);
    }

    public isInstanceNode(): boolean {
        return (NodeClass.Object === this.nodeClass ||
                NodeClass.Variable === this.nodeClass ||
                NodeClass.Method === this.nodeClass);
    }

    get name(): string { 
        return this._browseName; 
    }

    get browseName(): string {
        return this._browseName;
    }

    set browseName(browseName: string) {
        this._browseName = browseName;
    }

    get displayName(): UaLocalizedText {
        return this._displayName;
    }

    set displayName(displayName: UaLocalizedText) {
        this._displayName = displayName;
    }

    get writeMask(): number {
        return this._writeMask;
    }

    set writeMask(writeMask: number) {
        this._writeMask = writeMask;
    }

    get description(): UaLocalizedText {
        return this._description;
    }

    set description(description: UaLocalizedText) {
        this._description = description;
    }

    getAttribute(attributeId: number): UaVariant {
        let value: UaVariant = UaVariant.null();
        
        if (Attributes.NodeId === attributeId) {
            value = UaVariant.nodeId(this._nodeId);
        } else if (Attributes.NodeClass === attributeId) {
            value = UaVariant.integer(this.nodeClass, UaVariantType.Int32);
        } else if (Attributes.BrowseName === attributeId) {
            value = UaVariant.qualifiedName(this._browseName);
        } else if (Attributes.DisplayName === attributeId) {
            value = UaVariant.localizedText(this._displayName);
        } else if (Attributes.Description === attributeId) {
            if (this._description.text.length != 0) value = UaVariant.localizedText(this._description);            
        } else if (Attributes.WriteMask === attributeId) {
            value = UaVariant.integer(this._writeMask, UaVariantType.UInt32);
        } else if (Attributes.UserWriteMask === attributeId) {
            value = UaVariant.integer(0, UaVariantType.UInt32);
        }

        return value;
    }
    
    getReferences(direction: BrowseDirection): UaReference[] {
        const ret: UaReference[] = [];

        for (let item of this._references) {
            if ((BrowseDirection.Forward === direction && !item.isForward) ||
                (BrowseDirection.Inverse === direction && item.isForward)) continue;

            ret.push(item);
        }

        return ret;
    }

    addReference(reference: UaReference): void {
        this._references.push(reference);
    }
    
    getMembers(): UaInstanceNode[] {
        const referenceList = this.getReferences(BrowseDirection.Forward);
        const members: UaInstanceNode[] = [];

        for (const item of referenceList) {
            if (!item.linkedNode.isInstanceNode()) continue;
            if (!item.reference.isSubtypeOf(UaNodeId.from(ReferenceTypeIds.Aggregates))) continue;
            members.push(item.linkedNode as UaInstanceNode);
        }

        return members;
    }

    getMember(path: string): UaInstanceNode | null {
        for (const item of this._references) {
            if (item.isForward &&
                item.linkedNode.browseName === path &&
                item.linkedNode.isInstanceNode() &&
                item.reference.isSubtypeOf(UaNodeId.from(ReferenceTypeIds.Aggregates))) {
                return item.linkedNode as UaInstanceNode;
            }
        }
        return null;
    }

    protected addMemberNode(member: UaInstanceNode) {
        const UaReferenceTypes = getUaReferenceTypes();
        let reference = UaReferenceTypes.HasComponent;

        if (NodeClass.Variable == member.nodeClass) {
            if ((member as UaVariable).typeDefinition.nodeId.equal(UaNodeId.from(VariableTypeIds.PropertyType)))
            {
                reference = UaReferenceTypes.HasProperty;
            }
        }

        this.addReference(new UaReference(member, reference, true));
        member.addReference(new UaReference(this,reference, false));
    }
}
